packages/labs/analyzer/src/lib/javascript/classes.ts

```diff
@@ -308,31 +312,62 @@ export const getHeritage = (
 
 export const getHeritageFromExpression = (
   expression: ts.Expression,
-  analyzer: AnalyzerInterface
+  analyzer: AnalyzerInterface,
+  isMixinClass?: boolean
 ): ClassHeritage => {
-  // TODO(kschaaf): Support for extracting mixing applications from the heritage
-  // expression https://github.com/lit/lit/issues/2998
   const mixins: Reference[] = [];
-  const superClass = getSuperClass(expression, analyzer);
+  const superClass = getSuperClassAndMixins(expression, mixins, analyzer);
   return {
-    superClass,
+    superClass: isMixinClass ? undefined : superClass,
     mixins,
   };
 };
 
-export const getSuperClass = (
+export const getSuperClassAndMixins = (
   expression: ts.Expression,
+  foundMixins: Reference[],
   analyzer: AnalyzerInterface
 ): Reference | undefined => {
   // TODO(kschaaf) Could add support for inline class expressions here as well
   if (analyzer.typescript.isIdentifier(expression)) {
     return getReferenceForIdentifier(expression, analyzer);
+  } else if (
+    analyzer.typescript.isCallExpression(expression) &&
+    analyzer.typescript.isIdentifier(expression.expression)
+  ) {
+    // FYI we purposely restrict to identifiers since we represent mixins
+    // as references. If we want to support complex expressions in future,
+    // we will need to rework that in the model.
+    const mixinRef = getReferenceForIdentifier(expression.expression, analyzer);
+    // We need to eagerly dereference a mixin ref to know what argument the
+    // super class is passed into
+    const mixin = mixinRef?.dereference(MixinDeclaration);
+    // TODO (43081j): consider supporting external mixins properly at some point
+    if (mixinRef === undefined || mixin === undefined) {
+      analyzer.addDiagnostic(
+        createDiagnostic({
+          typescript: analyzer.typescript,
+          node: expression,
+          message:
+            `This is presumed to be a mixin but it could not be found ` +
+            `in the current project. Mixins imported from outside the ` +
+            `project are not yet supported.`,
```

Comment:
How were mixins handled prior to this change? Were they skipped or did they have a diagnostic. Wondering if the diagnostic should remain, but the analysis should try and continue with a graceful fallback of continuing to traverse the mixin expressions.

E.g. `const myClass = localMixin(externalMixin(BaseClass))` could maybe skip the `externalMixin` but still return `BaseClass` as the super class.

Or would we prefer to error than provide potentially wrong analysis results?

No action required - mostly a question so I can better understand the error handling philosophy of this package.

Reply:
If I'm following this correctly, we're going from TS expression -> model reference and bailing if we don't find the model reference. This makes some sense for things where we need analysis info outside of the type system (attributes, events, etc). But we should be able to support analysis of a mixins' basic class contributions even when they're not found in the analyzed model. We have to process the mixin type to find properties, etc. That seems valuable for elements that use mixins that are plain JS and not part of the CEM ecosystem.

Reply:
keep in mind this particular error has a `TODO` above it to implement support for external mixins (i.e. those we can't get a reference to, but know exist)

so do we want to do that in this PR or a separate one?

if i followed correctly, i think thats what you're asking for (the same as the todo)

Reply:
yeah, I think for our own reference it might be nice to note these two types of "external". I wouldn't mind seeing an open issue and have the error message link to it.

Reply:
i've added more info and an issue at least